api/src/deployment/deployment_unit.rs

```diff
@@ -0,0 +1,969 @@
+       
+       Ok(DeploymentUnitBuilder { 
+        stage: WithAppliedHooks {
+            app_name: self.stage.app_name,
+            configs: self.stage.configs,
+            service_companions: self.stage.service_companions,
+            app_companions: self.stage.app_companions,
+            templating_only_service_configs: self.stage.templating_only_service_configs,
+            image_infos: self.stage.image_infos,
+            strategies: strategies,
+        }
+    })
+   }
+}
+
+impl DeploymentUnitBuilder<WithAppliedHooks> {
+    pub fn build(self) -> DeploymentUnit {
+        DeploymentUnit{
+            app_name: self.stage.app_name,
+            _configs: self.stage.configs,
+            _service_companions: self.stage.service_companions,
+            _app_companions: self.stage.app_companions,
+            _templating_only_service_configs: self.stage.templating_only_service_configs,
+            _image_infos: self.stage.image_infos,
+            strategies: self.stage.strategies
+        }
+    }
+}
+
+
+fn resolve_strategy(
+    service_config: ServiceConfig,
+    strategy: &crate::config::DeploymentStrategy,
+    image_infos: &HashMap<Image, ImageInfo>,
+) -> DeploymentStrategy {
+    match strategy {
+        crate::config::DeploymentStrategy::RedeployAlways => {
+            DeploymentStrategy::RedeployAlways(service_config)
+        }
+        crate::config::DeploymentStrategy::RedeployOnImageUpdate => {
+            match image_infos.get(service_config.image()) {
+                Some(image_info) => DeploymentStrategy::RedeployOnImageUpdate(
+                    service_config,
+                    image_info.digest().to_string(),
+                ),
+                None => DeploymentStrategy::RedeployAlways(service_config),
+            }
+        }
+        crate::config::DeploymentStrategy::RedeployNever => {
+            DeploymentStrategy::RedeployNever(service_config)
+        }
+    }
+}
+
+fn container_type_index(container_type: &ContainerType) -> i32 {
+    match container_type {
+        ContainerType::ApplicationCompanion => 0,
+        ContainerType::ServiceCompanion => 1,
+        ContainerType::Instance | ContainerType::Replica => 2,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::models::{Environment, EnvironmentVariable};
+    use crate::{config_from_str, sc};
+    use secstr::SecUtf8;
+    use std::str::FromStr;
+
+
+    #[tokio::test]
+    async fn should_return_unique_images() -> Result<(),AppsServiceError>{
+        let config = Config::default();
+        let unit = DeploymentUnitBuilder::init(AppName::from_str("master").unwrap(), vec![sc!("http1", "nginx:1.13"), sc!("http2", "nginx:1.13")])
+            .extend_with_config(&config)
+            .extend_with_templating_only_service_configs(vec![sc!("http1", "nginx:1.13"), sc!("http2", "nginx:1.13")])
+            .images();
+        
+        assert_eq!(unit.len(), 1);
```

Comment:
Nitpick: I would test that unit (images would be a better name, BTW) contains "nginx:1.13" only.

Reply:
Yes, the test name states unique. I think we can pass 2 different services and check the uniqueness
vec![
                sc!("http1", "nginx:1.13"),
                sc!("wordpress1", "wordpress:alpine"),
            ]
  If i pass the above config then the test name - should_return_unique_images() will be the right name.